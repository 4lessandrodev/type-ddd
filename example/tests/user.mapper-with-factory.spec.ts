import {
	Model,
	AggregateFactory,
	ModelFactory,
} from '../user.mapper-with-factory';
import { User } from '../simple-user.aggregate';
import { UserMapper } from '../user.mapper-with-factory';
import {
	BirthdayValueObject,
	DomainId,
	EmailValueObject,
	PasswordValueObject,
	UserNameValueObject,
} from '@types-ddd';

const currentDate = new Date();
const lastYear = new Date(
	new Date().setFullYear(currentDate.getFullYear() - 1)
);

const model: Model = {
	id: 'valid_id',
	createdAt: currentDate,
	updatedAt: currentDate,
	userBirthDay: lastYear,
	userEmail: 'valid_email@domain.com',
	userName: 'Valid Name',
	userPassword: 'valid_pass',
};

const aggregate: User = User.create({
	ID: DomainId.create(model.id),
	userBirthDay: BirthdayValueObject.create(lastYear).getResult(),
	userEmail: EmailValueObject.create(model.userEmail).getResult(),
	userName: UserNameValueObject.create(model.userName).getResult(),
	userPassword: PasswordValueObject.create(model.userPassword).getResult(),
	createdAt: currentDate,
	updatedAt: currentDate,
}).getResult();

describe('user.factory', () => {
	it('should be defined', () => {
		const factory = AggregateFactory(model);
		expect(factory).toBeDefined();
	});

	it('should convert to model with success', () => {
		const result = AggregateFactory(model);
		const generatedAggregate = result.getResult();
		expect(result.isSuccess).toBeTruthy();
		expect(generatedAggregate).toEqual(aggregate);
		expect(generatedAggregate).toBeInstanceOf(User);
		expect(generatedAggregate.userEmail.value).toBe(model.userEmail);
		expect(generatedAggregate.id.value.toString()).toBe(model.id);
	});

	it('should return result.fail if provide an invalid value', () => {
		const invalidModel = Object.assign(
			{},
			{ ...model },
			{
				userEmail: 'invalid_email',
			}
		);
		const result = AggregateFactory(invalidModel);
		expect(result.isFailure).toBeTruthy();
	});
});

describe('model.factory', () => {
	it('should be defined', () => {
		const factory = ModelFactory(aggregate);
		expect(factory).toBeDefined();
	});

	it('should convert a aggregate into model with success', () => {
		const generatedModel = ModelFactory(aggregate);
		expect(generatedModel).toEqual(model);
		expect(generatedModel.id).toEqual(aggregate.id.value.toString());
		expect(generatedModel.userEmail).toEqual(aggregate.userEmail.value);
	});
});

describe('user.mapper-with-factory', () => {
	it('should be defined', () => {
		const mapper = new UserMapper(ModelFactory, AggregateFactory);
		expect(mapper).toBeDefined();
	});

	it('should convert from database to domain with success', () => {
		const mapper = new UserMapper(ModelFactory, AggregateFactory);
		const generatedAggregate = mapper.toDomain(model);
		expect(generatedAggregate).toEqual(aggregate);
		expect(generatedAggregate).toBeInstanceOf(User);
		expect(generatedAggregate.userEmail.value).toBe(model.userEmail);
		expect(generatedAggregate.id.value.toString()).toBe(model.id);
	});

	it('should convert from domain to database with success', () => {
		const mapper = new UserMapper(ModelFactory, AggregateFactory);
		const generatedModel = mapper.toPersistence(aggregate);
		expect(generatedModel).toEqual(model);
		expect(generatedModel.userEmail).toBe(aggregate.userEmail.value);
		expect(generatedModel.id).toBe(aggregate.id.value.toString());
	});
});
