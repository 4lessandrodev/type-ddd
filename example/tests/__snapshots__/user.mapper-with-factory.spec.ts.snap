// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user.mapper-with-factory should convert from database to domain with success 1`] = `
User {
  "_domainEvents": Array [],
  "_id": DomainId {
    "isNew": false,
    "props": Object {
      "value": "valid_id",
    },
  },
  "entityName": "User",
  "props": Object {
    "ID": DomainId {
      "isNew": false,
      "props": Object {
        "value": "valid_id",
      },
    },
    "createdAt": 2022-01-01T03:00:00.000Z,
    "updatedAt": 2022-01-01T03:00:00.000Z,
    "userBirthDay": BirthdayValueObject {
      "ONE_YEAR": 31536000902,
      "props": Object {
        "value": 2022-01-01T03:00:00.000Z,
      },
    },
    "userEmail": EmailValueObject {
      "props": Object {
        "value": "valid_email@domain.com",
      },
    },
    "userName": UserNameValueObject {
      "props": Object {
        "value": "Valid Name",
      },
    },
    "userPassword": PasswordValueObject {
      "props": Object {
        "value": "valid_pass",
      },
    },
  },
}
`;

exports[`user.mapper-with-factory should convert from domain to database with success 1`] = `
Object {
  "createdAt": 2022-01-01T03:00:00.000Z,
  "id": "valid_id",
  "updatedAt": 2022-01-01T03:00:00.000Z,
  "userBirthDay": 2022-01-01T03:00:00.000Z,
  "userEmail": "valid_email@domain.com",
  "userName": "Valid Name",
  "userPassword": "valid_pass",
}
`;
